### PR TITLE
Refactor get xcos

### PR DESCRIPTION
**Pull request by Nash2325138**
- Add customized dataset classes for IJB-A/C and ARFace
- Add the notebooks to run and evaluate scores on the original model and our xCos model
- Other scripts like ARFace dataset alignment (by MTCNN)

### CODE REVIEW

File: data/data_pipe.py [hunk 1]
```diff
@@ -108,6 +116,37 @@ def load_mx_rec(rec_path):
...
+            output = torch.tensor(self.carray[self.idx: self.idx + self.batch_size])
+        elif self.idx < len(self.carray):
+            output = torch.tensor(self.carray[self.idx:])
+        else:
+            raise StopIteration
+        self.idx += self.batch_size
+        return output
+
+
+def loader_from_carray(carray: np.ndarray, conf, strategy='iterable'):
```
Comment (ntubiolin) [inline data/data_pipe.py:140]: When will we want to use the strategy='iterable'?

File: data/datasets.py [hunk 2]
```diff
@@ -750,27 +750,27 @@ def __init__(self, ijbc_data_root):
         self.match = pd.read_csv(op.join(test1_dir, 'match.csv'))
 
         self.transforms = transforms.Compose([
-            transforms.Normalize([.5, .5, .5], [.5, .5, .5]),
             transforms.Resize([112, 112]),
-            transforms.toTensor()
+            transforms.ToTensor(),
+            transforms.Normalize([.5, .5, .5], [.5, .5, .5]),
```
Comment (ntubiolin) [inline data/datasets.py]: Maybe this is the reason why the performance degrades

File: mtcnn_pytorch/src/align_trans.py [hunk 3]
```diff
@@ -246,10 +246,10 @@ def warp_and_crop_face(src_img,
     """
 
     if reference_pts is None:
-        if crop_size[0] == 96 and crop_size[1] == 112:
+        if crop_size is not None and crop_size[0] == 96 and crop_size[1] == 112:
```
Comment (ntubiolin) [inline mtcnn_pytorch/src/align_trans.py:249]: When will we meet the situation that crop_size is not None?

File: mtcnn_pytorch/src/align_trans.py [hunk 4]
```diff
@@ -246,10 +246,10 @@ def warp_and_crop_face(src_img,
...
             reference_pts = REFERENCE_FACIAL_POINTS
         else:
-            default_square = False
+            default_square = True
```
Comment (ntubiolin) [inline mtcnn_pytorch/src/align_trans.py:252]: Why?

File: verification.py [hunk 5]
```diff
@@ -160,27 +160,24 @@ def calculate_accuracy(threshold, dist, actual_issame, useCos=False):
...
-    :param far_target:
-    :param nrof_folds:
-    :return:
-    '''
     assert (embeddings1.shape[0] == embeddings2.shape[0])
     assert (embeddings1.shape[1] == embeddings2.shape[1])
-    nrof_pairs = min(len(actual_issame), embeddings1.shape[0])
+    diff = np.subtract(embeddings1, embeddings2)
+    dist = np.sum(np.square(diff), 1)
+    return calculate_val(thresholds, dist, actual_issame, far_target, nrof_folds=nrof_folds)
```
Comment (ntubiolin) [inline verification.py]: Should this line be 
```suggestion
    return calculate_val_by_diff(thresholds, dist, actual_issame, far_target, nrof_folds=nrof_folds)
```
?

**Review by ntubiolin — COMMENTED** [review]
Please review those comments.

File: data/data_pipe.py (hunk 1 above)
Comment (Nash2325138) [reply data/data_pipe.py:140]: When the whole dataset is small enough to be put in the memory

File: data/datasets.py (hunk 2 above)
Comment (Nash2325138) [reply data/datasets.py]: Why?

File: mtcnn_pytorch/src/align_trans.py (hunk 3 above)
Comment (Nash2325138) [reply mtcnn_pytorch/src/align_trans.py:249]: You mean when will it be None?

File: mtcnn_pytorch/src/align_trans.py (hunk 4 above)
Comment (Nash2325138) [reply mtcnn_pytorch/src/align_trans.py:252]: To avoid some bugs, but I've forgot the actual behaviour of that bug.

**Pull request merged**